apps/labs/posts/scikit-lego-narwhals.md

```diff
@@ -0,0 +1,115 @@
+
+- once the input has been converted to pandas, there's no standardised way of converting back to
+  the original dataframe library.
+- using pandas as an engine may be significantly slower than doing everything using the original
+  dataframe
+
+Furthermore, converting to pandas may present a cost - for example, if you start with a Polars
+LazyFrame, which can run significantly faster because it delays reading data into memory and performs
+[several optimisations](https://docs.pola.rs/user-guide/lazy/optimizations/),
+then you're required to call [`.collect`](https://docs.pola.rs/py-polars/html/reference/lazyframe/api/polars.LazyFrame.collect.html) on it before converting to pandas.
+
+To drive the second point home, here are some timings comparing:
+
+- running `add_lags` on a 10-million-row Polars dataframe directly
+- converting that same dataframe to pandas and running `add_lags` on that
+
+![with native Polars support `add_lags` takes ~ 11 ms, whereas going via pandas it
+takes ~ 1470 ms. Image produced with `great-tables` package](/posts/scikit-lego-narwhals/comparison.png)
+
+[Code to reproduce](https://gist.github.com/MarcoGorelli/1da1971063caf0b3e5133f5dfba3315b).
+
+Note that we're dealing with purely numeric data here, so the conversion is extremely cheap.
+The reason that the second approach takes over 100x longer is that Polars generally excels at
+feature engineering. As the size of the dataframe increases, the comparison
+scales even better in the "native Polars support" direction.
+
+That's why "just convert to pandas" is a disappointing solution, and why it's worth it to
+invest the time to achieve dataframe-agnosticism.
+
+## Will the future of data science become "BYODF" (bring your own dataframe)?
+
+Earlier this year, we saw [Great Tables announce that they are now "bring your own dataframe"](https://posit-dev.github.io/great-tables/blog/bring-your-own-df/).
+Scikit-learn also supports Polars in a somewhat standardised way, by leveraging the [dataframe interchange protocol](https://data-apis.org/dataframe-protocol/latest/), the [`__array__` method](https://numpy.org/devdocs/user/basics.interoperability.html), and
+some custom logic. And in this blog post, we look at how scikit-lego achieved the same by using Narwhals.
+
+Is this a sign of what's to come? I certainly hope so. pandas is - and will very likely continue to be -
+an enormously useful tool which solves a lot of problems for a lot of real people. 
+However, if pandas remained the _only_ dataframe library supported in the data science
+ecosystem, then that'd be a missed opportunity.
+
+We've looked at how scikit-lego leveraged Narwhals in order to become dataframe-agnostic,
+able to natively support Polars, Modin, and cuDF without adding them as dependencies.
+Let's hope that this is the direction that data science libraries are headed towards, and that Narwhals
+can facilitate the process.
```

Comment:
The concluding paragraphs kind of do an odd loop/return here. I feel like you could make this concluding section stronger, but it's also not bad. I edited the last paragraph to remove some of the echo effect and for brevity.

```suggestion
Scikit-lego leveraged Narwhals in order to become dataframe-agnostic,
able to natively support Polars, Modin, and cuDF without adding them as dependencies.
Let's hope that other data science libraries do the same.
```

Reply:
i wasn't familiar with the term "echo effect", but I recognise it when I see it - thanks!

Reply:
[I make up terms all the time](https://github.com/Quansight-Labs/czi-scientific-python-mgmt/issues/88)